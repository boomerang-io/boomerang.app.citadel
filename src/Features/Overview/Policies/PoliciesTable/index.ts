// @ts-expect-error ts-migrate(6142) FIXME: Module './PoliciesTable' was resolved to 'C:/Users... Remove this comment to see the full error message
export { default } from "./PoliciesTable"