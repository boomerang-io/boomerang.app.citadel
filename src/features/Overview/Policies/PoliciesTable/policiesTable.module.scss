@import "~styles/variables";

.tableContainer {
  margin-top: 1rem;
}

.tableHeadRow {
  //background-color: $cool-white-1;
}

.tableHeadRow > .tableHeader {
  color: $interactive-01;
  font-size: inherit;
  font-family: inherit;
  font-weight: 600;
  height: 3rem;

  &.policyName {
    width: 40%;
  }

  &.definitions {
    width: 20%;
  }

  &.rules {
    width: 20%;
  }

  &.createdDate {
    width: 20%;
  }
}
.tableBody {
  background-color: #f2f2f2;
  color: gray;
  font-weight: 400;
}
.tableRow {
  cursor: pointer;
}
.tableCell {
  align-items: center;
  display: flex;
  height: 3.5rem;
  padding-left: 0.75rem;
  width: 100%;
}
.tableIcons {
  display: flex;
  justify-content: flex-end;
  width: 100%;
}
.tableIcon {
  cursor: pointer;
  display: flex;
  height: 1.25rem;
  margin-right: 1.5rem;
  width: 1.25rem;
}

.tableTextarea {
  background: inherit;
  color: $text-01;
  max-width: 100%;
  resize: none;
  text-align: left;
  white-space: normal;
  width: 90%;
  margin-left: 0.25rem;
}
