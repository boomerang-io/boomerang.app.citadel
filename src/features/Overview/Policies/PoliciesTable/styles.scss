@import "~@boomerang/boomerang-styles/lib/colors";
@import "~Styles/variables";

#policies-container {
  .bx--table-header-label {
    color: $cerulean-50;
    font-size: 0.9rem;
    font-weight: bolder;
    padding-right: 0.5rem;
  }
}