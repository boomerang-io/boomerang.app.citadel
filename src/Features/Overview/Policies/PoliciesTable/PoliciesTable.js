import React from "react";
import PropTypes from "prop-types";
import { useHistory, useParams } from "react-router-dom";
import moment from "moment";
import { DataTable } from "@boomerang-io/carbon-addons-boomerang-react";
import { appLink } from "Config/appConfig";
import styles from "./policiesTable.module.scss";

const headers = [
  {
    header: "Name",
    key: "name",
  },

  {
    header: "Definitions",
    key: "definitions",
  },
  {
    header: "Rules",
    key: "rules",
  },
  {
    header: "Created Date",
    key: "createdDate",
  },
];

PoliciesTable.propTypes = {
  poicies: PropTypes.array,
  definitions: PropTypes.array,
};

export default function PoliciesTable(props) {
  let history = useHistory();
  let params = useParams();
  const handleRowClick = (row) => {
    history.push(appLink.editPolicy({ teamId: params.teamId, policyId: row.id }));
  };

  const renderCell = (cells, cellIndex, value) => {
    const column = headers[cellIndex];
    switch (column.header) {
      case "Name":
        return <p className={styles.tableTextarea}>{value}</p>;
      case "Definitions":
        return <p className={styles.tableTextarea}>{Array.isArray(value) ? value.length : "---"}</p>;
      case "Created Date":
        return <time className={styles.tableTextarea}>{value ? moment(value).format("MMMM DD, YYYY") : "---"}</time>;
      case "Rules":
        const defValue = cells.find((cell) => cell.id.includes("definitions")).value;
        return (
          <p className={styles.tableTextarea}>
            {Array.isArray(defValue)
              ? defValue.reduce((accum, definition) => {
                  accum += definition.rules.length;
                  return accum;
                }, 0)
              : "---"}
          </p>
        );

      default:
        return value || "---";
    }
  };

  const { policies } = props;
  const { TableContainer, Table, TableHead, TableRow, TableBody, TableCell, TableHeader } = DataTable;
  return (
    <DataTable
      rows={policies}
      headers={headers}
      isSortable={true}
      render={({ rows, headers, getHeaderProps }) => (
        <TableContainer>
          <Table isSortable className={styles.tableContainer} useZebraStyles={false}>
            <TableHead>
              <TableRow className={styles.tableHeadRow}>
                {headers.map((header) => (
                  <TableHeader
                    {...getHeaderProps({ header, className: `${styles.tableHeader} ${styles[header.key]}` })}
                  >
                    {header.header}
                  </TableHeader>
                ))}
              </TableRow>
            </TableHead>
            <TableBody className={styles.tableBody} data-testid="policies-tbody">
              {rows.map((row, rowIndex) => (
                <TableRow
                  key={row.id}
                  className={styles.tableRow}
                  onClick={() => handleRowClick(row)}
                  data-testid="policies-table-row"
                >
                  {row.cells.map((cell, cellIndex) => (
                    <TableCell key={`${cell.id}-${cellIndex}`} style={{ padding: "0" }}>
                      <div className={styles.tableCell}>{renderCell(row.cells, cellIndex, cell.value)}</div>
                    </TableCell>
                  ))}
                </TableRow>
              ))}
            </TableBody>
          </Table>
        </TableContainer>
      )}
    />
  );
}
