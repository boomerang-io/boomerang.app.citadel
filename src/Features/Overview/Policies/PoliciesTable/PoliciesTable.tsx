import React from "react";
// @ts-expect-error ts-migrate(7016) FIXME: Could not find a declaration file for module 'reac... Remove this comment to see the full error message
import { useHistory, useParams } from "react-router-dom";
import moment from "moment";
// @ts-expect-error ts-migrate(7016) FIXME: Could not find a declaration file for module '@boo... Remove this comment to see the full error message
import { DataTable } from "@boomerang-io/carbon-addons-boomerang-react";
// @ts-expect-error ts-migrate(2307) FIXME: Cannot find module 'Config/appConfig' or its corre... Remove this comment to see the full error message
import { appLink } from "Config/appConfig";
// @ts-expect-error ts-migrate(2307) FIXME: Cannot find module './policiesTable.module.scss' o... Remove this comment to see the full error message
import styles from "./policiesTable.module.scss";

const headers = [
  {
    header: "Name",
    key: "name",
  },

  {
    header: "Definitions",
    key: "definitions",
  },
  {
    header: "Rules",
    key: "rules",
  },
  {
    header: "Created Date",
    key: "createdDate",
  },
];

type Props = {
    poicies?: any[];
    definitions?: any[];
};

export default function PoliciesTable(props: Props) {
  let history = useHistory();
  let params = useParams();
  const handleRowClick = (row: any) => {
    history.push(appLink.editPolicy({ teamId: params.teamId, policyId: row.id }));
  };

  const renderCell = (cells: any, cellIndex: any, value: any) => {
    const column = headers[cellIndex];
    switch (column.header) {
      case "Name":
        // @ts-expect-error ts-migrate(17004) FIXME: Cannot use JSX unless the '--jsx' flag is provided... Remove this comment to see the full error message
        return <p className={styles.tableTextarea}>{value}</p>;
      case "Definitions":
        // @ts-expect-error ts-migrate(17004) FIXME: Cannot use JSX unless the '--jsx' flag is provided... Remove this comment to see the full error message
        return <p className={styles.tableTextarea}>{Array.isArray(value) ? value.length : "---"}</p>;
      case "Created Date":
        // @ts-expect-error ts-migrate(17004) FIXME: Cannot use JSX unless the '--jsx' flag is provided... Remove this comment to see the full error message
        return <time className={styles.tableTextarea}>{value ? moment(value).format("MMMM DD, YYYY") : "---"}</time>;
      case "Rules":
        const defValue = cells.find((cell: any) => cell.id.includes("definitions")).value;
        return (
          // @ts-expect-error ts-migrate(17004) FIXME: Cannot use JSX unless the '--jsx' flag is provided... Remove this comment to see the full error message
          <p className={styles.tableTextarea}>
            {Array.isArray(defValue)
              ? defValue.reduce((accum, definition) => {
                  accum += definition.rules.length;
                  return accum;
                }, 0)
              : "---"}
          </p>
        );

      default:
        return value || "---";
    }
  };

  // @ts-expect-error ts-migrate(2339) FIXME: Property 'policies' does not exist on type 'Props'... Remove this comment to see the full error message
  const { policies } = props;
  const { TableContainer, Table, TableHead, TableRow, TableBody, TableCell, TableHeader } = DataTable;
  return (
    // @ts-expect-error ts-migrate(17004) FIXME: Cannot use JSX unless the '--jsx' flag is provided... Remove this comment to see the full error message
    <DataTable
      rows={policies}
      headers={headers}
      isSortable={true}
      render={({
        rows,
        headers,
        getHeaderProps
      }: any) => (
        // @ts-expect-error ts-migrate(17004) FIXME: Cannot use JSX unless the '--jsx' flag is provided... Remove this comment to see the full error message
        <TableContainer>
          {/* @ts-expect-error ts-migrate(17004) FIXME: Cannot use JSX unless the '--jsx' flag is provided... Remove this comment to see the full error message */}
          <Table isSortable className={styles.tableContainer} useZebraStyles={false}>
            {/* @ts-expect-error ts-migrate(17004) FIXME: Cannot use JSX unless the '--jsx' flag is provided... Remove this comment to see the full error message */}
            <TableHead>
              {/* @ts-expect-error ts-migrate(17004) FIXME: Cannot use JSX unless the '--jsx' flag is provided... Remove this comment to see the full error message */}
              <TableRow className={styles.tableHeadRow}>
                {/* @ts-expect-error ts-migrate(17004) FIXME: Cannot use JSX unless the '--jsx' flag is provided... Remove this comment to see the full error message */}
                {headers.map((header: any) => <TableHeader
                  {...getHeaderProps({ header, className: `${styles.tableHeader} ${styles[header.key]}` })}
                >
                  {header.header}
                </TableHeader>)}
              </TableRow>
            </TableHead>
            {/* @ts-expect-error ts-migrate(17004) FIXME: Cannot use JSX unless the '--jsx' flag is provided... Remove this comment to see the full error message */}
            <TableBody className={styles.tableBody} data-testid="policies-tbody">
              {rows.map((row: any, rowIndex: any) => (
                // @ts-expect-error ts-migrate(17004) FIXME: Cannot use JSX unless the '--jsx' flag is provided... Remove this comment to see the full error message
                <TableRow
                  key={row.id}
                  className={styles.tableRow}
                  onClick={() => handleRowClick(row)}
                  data-testid="policies-table-row"
                >
                  {row.cells.map((cell: any, cellIndex: any) => (
                    // @ts-expect-error ts-migrate(17004) FIXME: Cannot use JSX unless the '--jsx' flag is provided... Remove this comment to see the full error message
                    <TableCell key={`${cell.id}-${cellIndex}`} style={{ padding: "0" }}>
                      {/* @ts-expect-error ts-migrate(17004) FIXME: Cannot use JSX unless the '--jsx' flag is provided... Remove this comment to see the full error message */}
                      <div className={styles.tableCell}>{renderCell(row.cells, cellIndex, cell.value)}</div>
                    </TableCell>
                  ))}
                </TableRow>
              ))}
            </TableBody>
          </Table>
        </TableContainer>
      )}
    />
  );
}
