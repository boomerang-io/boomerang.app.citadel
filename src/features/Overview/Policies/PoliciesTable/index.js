import React, { Component } from "react";
import PropTypes from "prop-types";
import { withRouter } from "react-router-dom";
import { DataTable } from "carbon-components-react";
import styles from "./policiesTable.module.scss";
import "./styles.scss";

export class PoliciesTable extends Component {
  static propTypes = {
    poicies: PropTypes.array,
    definitions: PropTypes.array
  };
  headers = [
    {
      header: "Name",
      key: "name"
    },

    {
      header: "Definitions",
      key: "definitions"
    },
    {
      header: "Rules",
      key: "definitions"
    },
    {
      header: "Stage Gate Allocations",
      key: "stages"
    }
  ];

  handleRowClick = row => {
    const { location, history } = this.props;
    history.push(`${location.pathname}/policy/edit/${row.id}`);
  };

  renderCell = (rowIndex, cellIndex, value) => {
    const column = this.headers[cellIndex];
    switch (column.header) {
      case "Name":
        return <p className={styles.tableTextarea}>{value}</p>;
      case "Definitions":
        return <p className={styles.tableTextarea}>{Array.isArray(value) ? value.length : "---"}</p>;
      case "Rules":
        return (
          <p className={styles.tableTextarea}>
            {Array.isArray(value)
              ? value.reduce((accum, definition) => {
                  accum += definition.rules.length;
                  return accum;
                }, 0)
              : "---"}
          </p>
        );

      case "Stage Gate Allocations":
        return <p className={styles.tableTextarea}>{!!value.length ? value.join(", ") : "---"}</p>;
      case "":
        return (
          <div className={styles.tableIcons}>
            {/* <div>
              <Icon
                data-tip
                data-for={`policies-table-${value}-${cellIndex}-edit`}
                className={styles.tableIcon}
                name="icon--edit"
                alt="Edit Policy"
              />
              <Tooltip id={`policies-table-${value}-${cellIndex}-edit`}>Edit</Tooltip>
            </div>
            <div>
              <Icon
                data-tip
                data-for={`policies-table-${value}-${cellIndex}-delete`}
                className={styles.tableIcon}
                name="icon--delete"
                alt="Delete Policy"
              />
              <Tooltip id={`policies-table-${value}-${cellIndex}-delete`}>Delete</Tooltip>
            </div> */}
          </div>
        );
      default:
        return value || "---";
    }
  };

  render() {
    const { policies } = this.props;
    const { TableContainer, Table, TableHead, TableRow, TableBody, TableCell, TableHeader } = DataTable;
    return (
      <DataTable
        rows={policies}
        headers={this.headers}
        render={({ rows, headers, getHeaderProps }) => (
          <TableContainer>
            <Table className={styles.tableContainer} sortable={true} zebra={false}>
              <TableHead>
                <TableRow className={styles.tableHeadRow}>
                  {headers.map(header => (
                    <TableHeader {...getHeaderProps({ header, className: `${styles.tableHeader}` })}>
                      {header.header}
                    </TableHeader>
                  ))}
                </TableRow>
              </TableHead>
              <TableBody className={styles.tableBody} data-testid="policies-tbody">
                {rows.map((row, rowIndex) => (
                  <TableRow key={row.id} className={styles.tableRow} onClick={() => this.handleRowClick(row)}>
                    {row.cells.map((cell, cellIndex) => (
                      <TableCell key={`${cell.id}-${cellIndex}`} style={{ padding: "0" }}>
                        <div className={styles.tableCell}>{this.renderCell(rowIndex, cellIndex, cell.value)}</div>
                      </TableCell>
                    ))}
                  </TableRow>
                ))}
              </TableBody>
            </Table>
          </TableContainer>
        )}
      />
    );
  }
}

export default withRouter(PoliciesTable);
