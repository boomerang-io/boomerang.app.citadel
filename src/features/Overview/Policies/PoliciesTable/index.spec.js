import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import { MemoryRouter } from "react-router";
import PoliciesTable from "./index";

const policies = [
  {
    createdDate: 1558051200000,
    name: "Timss",
    teamId: "5a8b331e262a70306622df73",
    definitions: [
      {
        policyDefinitionId: "5cd49777f6ea74a9bb6ac629",
        rules: [
          {
            value: "sdfgsdfg",
            operator: "less than",
            metric: "complexity"
          }
        ]
      },
      {
        policyDefinitionId: "5cd498f3f6ea74a9bb6ad0f3",
        rules: [
          {
            version: " asdfasdf",
            artifact: "asdasdf`",
            type: "maven"
          }
        ]
      }
    ],
    stages: ["dev"],
    id: "R4nuqJE"
  }
];

describe("PoliciesTable --- Snapshot", () => {
  it("Capturing Snapshot of PoliciesTable", () => {
    const renderedValue = renderer
      .create(
        <MemoryRouter>
          <PoliciesTable policies={policies} />
        </MemoryRouter>
      )
      .toJSON();
    expect(renderedValue).toMatchSnapshot();
  });
});

describe("PoliciesTable --- Shallow render", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <MemoryRouter>
        <PoliciesTable policies={policies} />
      </MemoryRouter>
    );
  });

  it("Render the DUMB component", () => {
    expect(wrapper.length).toEqual(1);
  });
});
