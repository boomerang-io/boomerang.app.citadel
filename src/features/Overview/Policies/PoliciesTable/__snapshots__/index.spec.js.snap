// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PoliciesTable --- Snapshot Capturing Snapshot of PoliciesTable 1`] = `
<div
  className="bx--data-table-v2-container"
>
  <table
    className="bx--data-table-v2 tableContainer"
    sortable={true}
  >
    <thead>
      <tr
        className="tableHeadRow"
      >
        <th
          aria-sort="none"
          className="tableHeader"
          scope="col"
        >
          <button
            className="tableHeader bx--table-sort-v2"
            onClick={[Function]}
          >
            <span
              className="bx--table-header-label"
            >
              Name
            </span>
            <svg
              alt="Sort rows by this header in ascending order"
              aria-label="Sort rows by this header in ascending order"
              className="bx--table-sort-v2__icon"
              fillRule="evenodd"
              height="5"
              role="img"
              viewBox="0 0 10 5"
              width="10"
            >
              <title>
                Sort rows by this header in ascending order
              </title>
              <path
                d="M0 5L5 .002 10 5z"
              />
            </svg>
          </button>
        </th>
        <th
          aria-sort="none"
          className="tableHeader"
          scope="col"
        >
          <button
            className="tableHeader bx--table-sort-v2"
            onClick={[Function]}
          >
            <span
              className="bx--table-header-label"
            >
              Definitions
            </span>
            <svg
              alt="Sort rows by this header in ascending order"
              aria-label="Sort rows by this header in ascending order"
              className="bx--table-sort-v2__icon"
              fillRule="evenodd"
              height="5"
              role="img"
              viewBox="0 0 10 5"
              width="10"
            >
              <title>
                Sort rows by this header in ascending order
              </title>
              <path
                d="M0 5L5 .002 10 5z"
              />
            </svg>
          </button>
        </th>
        <th
          aria-sort="none"
          className="tableHeader"
          scope="col"
        >
          <button
            className="tableHeader bx--table-sort-v2"
            onClick={[Function]}
          >
            <span
              className="bx--table-header-label"
            >
              Rules
            </span>
            <svg
              alt="Sort rows by this header in ascending order"
              aria-label="Sort rows by this header in ascending order"
              className="bx--table-sort-v2__icon"
              fillRule="evenodd"
              height="5"
              role="img"
              viewBox="0 0 10 5"
              width="10"
            >
              <title>
                Sort rows by this header in ascending order
              </title>
              <path
                d="M0 5L5 .002 10 5z"
              />
            </svg>
          </button>
        </th>
        <th
          aria-sort="none"
          className="tableHeader"
          scope="col"
        >
          <button
            className="tableHeader bx--table-sort-v2"
            onClick={[Function]}
          >
            <span
              className="bx--table-header-label"
            >
              Stage Gate Allocations
            </span>
            <svg
              alt="Sort rows by this header in ascending order"
              aria-label="Sort rows by this header in ascending order"
              className="bx--table-sort-v2__icon"
              fillRule="evenodd"
              height="5"
              role="img"
              viewBox="0 0 10 5"
              width="10"
            >
              <title>
                Sort rows by this header in ascending order
              </title>
              <path
                d="M0 5L5 .002 10 5z"
              />
            </svg>
          </button>
        </th>
        <th
          aria-sort="none"
          className="tableHeader"
          scope="col"
        >
          <button
            className="tableHeader bx--table-sort-v2"
            onClick={[Function]}
          >
            <span
              className="bx--table-header-label"
            >
              Violations
            </span>
            <svg
              alt="Sort rows by this header in ascending order"
              aria-label="Sort rows by this header in ascending order"
              className="bx--table-sort-v2__icon"
              fillRule="evenodd"
              height="5"
              role="img"
              viewBox="0 0 10 5"
              width="10"
            >
              <title>
                Sort rows by this header in ascending order
              </title>
              <path
                d="M0 5L5 .002 10 5z"
              />
            </svg>
          </button>
        </th>
      </tr>
    </thead>
    <tbody
      className="tableBody"
      data-testid="policies-tbody"
    >
      <tr
        className="tableRow"
        onClick={[Function]}
      >
        <td
          style={
            Object {
              "padding": "0",
            }
          }
        >
          <div
            className="tableCell"
          >
            <p
              className="tableTextarea"
            >
              Timss
            </p>
          </div>
        </td>
        <td
          style={
            Object {
              "padding": "0",
            }
          }
        >
          <div
            className="tableCell"
          >
            <p
              className="tableTextarea"
            >
              2
            </p>
          </div>
        </td>
        <td
          style={
            Object {
              "padding": "0",
            }
          }
        >
          <div
            className="tableCell"
          >
            <p
              className="tableTextarea"
            >
              2
            </p>
          </div>
        </td>
        <td
          style={
            Object {
              "padding": "0",
            }
          }
        >
          <div
            className="tableCell"
          >
            <p
              className="tableTextarea"
            >
              dev
            </p>
          </div>
        </td>
        <td
          style={
            Object {
              "padding": "0",
            }
          }
        >
          <div
            className="tableCell"
          >
            ---
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
